### PR TITLE
Fix parser for root (only) dataset names

### PR DESCRIPTION
**Pull request by jimklimov**
Includes PRs #619 (see for most context of the initial problem) and #620 (so also addresses spellchecker passing after these changes).

After some tinkering with the splitter logic vs. conjured-up test cases, just one case remained where we can only define our preference (at least, lacking some other method arguments or wholly a new class to convey the information explicitly): how to treat `X@Y:Z` strings - are they a remote `user@host:rootds` or a local `rootds@funny:snapname`?

So the self-test defines this one use-case as a "bogus" one with somewhat reasonably "hard-coded" specific expectations:

> NOTE: In absence of any hints we can not reliably discern below
> a     task='poolrootfs@snap-2:3'
> vs.   task='username@hostname:poolrootfs'
> which one is a local pool's root dataset with a funny but legal
> snapshot name, and which one is a remote user@host spec with a
> remote pool's root dataset. For practical purposes, we proclaim
> preference for the former: we are more likely to look at funny
> local snapshot names, than to back up to (or otherwise care about)
> remote pools' ROOT datasets.

...and so the self-tests for the parser are enabled as part of standard test runs here.

Also this uncovered deficiencies in `ZnapZend::Config` variant of the parser which got fixed now by the regex proposed in #585 (and improved upon here for `ZnapZend::ZFS`). It does not have to care about snapshots, so is a lot simpler.

### CODE REVIEW

**Comment by oetiker** [comment]
how about assuming case user@host:destds unless user actually is a local rootds ?

**Comment by jimklimov** [comment]
> > how about assuming case user@host:destds unless user actually is a local rootds ?

>  For practical purposes, we proclaim preference for the former: we are more likely to look at funny local snapshot names, than to back up to (or otherwise care about) remote pools' ROOT datasets.

Yep, so we declare that `user@host:destds` means local pool named `user` (and its root dataset), and its snapshot `host:destds`.

**Comment by oetiker** [comment]
what I mean is that the code could actually test(!) if a given string actually is a local root dataset

**Comment by jimklimov** [comment]
Ah, must have misread your earlier comment. Yes, I've had such thoughts too - just did not get to implementing them :)

Instead, wondered whether:

* reliability:
  * do we have some context that such pool does not exist currently so we can not reliably check if it is a pool?
    * e.g. a pluggable HDD/flash seems likely; rotation of several normally-vaulted media devices is a good practice for redundancy
    * pre-emptive setup to create a pool (more so via znapzend) seems unlikely
  * can then check if the `host` part can be resolved, but for systems with intermittent networking (VPN?) this may also not be bullet-proof
* a chicken-and-egg problem:
  * our generic ZFS calls IIRC use this helper method to decide if they should go over SSH :)
  * should we go for custom call-out to a local `zfs` tool and/or special tweaks to the common method?

**Comment by oetiker** [comment]
you are right this might be fragile ... another idea is this:

we know the actual src and destination configurations ... they do not include snapshots, so at that time things are still clear ... 

so I guess we just need to alter the parsing flow

**Comment by jimklimov** [comment]
> I guess we just need to alter the parsing flow

As noted earlier, for the least intrusive change, I am currently inclined in favor of further arguments (caller knows if the context deals with possible snapshots or strictly "live" dataset names).

Not sure if such refinement has to be part of this PR, or if you deem this one as an improvement (less-broken state of codebase) and it can be merged as a stepping-block to someone (else) making and diligently testing that change. Coding the change does not feel that hard, but testing and chizeling would likely need more time than I could currently share.

**Comment by jimklimov** [comment]
It may also be worthwhile to use `ZnapZend::Config` definitions of the splitting method for strings coming from the config, as opposed to those generally juggled  by `ZnapZend::ZFS`. OOP-wise, these strings reflect sort of different object classes. But other than such an idea, I have not much more about implementing something of the sort.

**Review by oetiker — APPROVED** [review]

**Review by oetiker — CHANGES_REQUESTED** [review]
please update the CHANGES file

**Comment by jimklimov** [comment]
Cheers, is the update okay? :)

**Review by oetiker — APPROVED** [review]

File: .github/workflows/spelling/expect.txt [hunk 1]
```diff
@@ -383,6 +383,7 @@ notest
 NOTMAKE
 nroff
 NSTIPS
+nsub
```
Comment (Ariessanea1) [inline .github/workflows/spelling/expect.txt:386]: Submit

Comment (Ariessanea1) [reply .github/workflows/spelling/expect.txt:386]: > Submit

File: .github/workflows/spelling/expect.txt [hunk 2]
```diff
@@ -383,6 +383,7 @@ notest
...
 nvpool
```
Comment (Ariessanea1) [inline .github/workflows/spelling/expect.txt:387]: Pool

File: lib/ZnapZend/Config.pm [hunk 3]
```diff
@@ -47,7 +47,9 @@ has backupSets => sub { [] };
 
 ### private functions ###
 my $splitHostDataSet = sub {
```
Comment (Ariessanea1) [inline lib/ZnapZend/Config.pm:49]: Set host bate

File: t/znapzend-lib-splitter.t [hunk 4]
```diff
@@ -4,7 +4,7 @@
 # since there are many use-cases and combinations to take care of.
 # We do so below by constructing "task" strings from components we
 # know to be a dataset name and some defined (or not) remote spec
-# and/or snapshot name, and deconstructing it back with the class
+# and/or snapshot name, and de-constructing it back with the class
```
Comment (Ariessanea1) [inline t/znapzend-lib-splitter.t:7]: Congratulations

File: .github/workflows/spelling/expect.txt [hunk 5]
```diff
@@ -383,6 +383,7 @@ notest
 NOTMAKE
 nroff
 NSTIPS
```
Comment (Ariessanea1) [inline .github/workflows/spelling/expect.txt:385]: Stops

File: .github/workflows/spelling/expect.txt [hunk 6]
```diff
@@ -419,6 +420,7 @@ plaintar
 plist
 plugin
 png
```
Comment (Ariessanea1) [inline .github/workflows/spelling/expect.txt:422]: Phnom

File: .github/workflows/spelling/expect.txt [hunk 7]
```diff
@@ -461,6 +463,7 @@ rmdir
 rmprog
 rmtmp
 ron
+rootds
```
Comment (Ariessanea1) [inline .github/workflows/spelling/expect.txt:466]: Roots

File: lib/ZnapZend/Config.pm (hunk 3 above)
Comment (Ariessanea1) [inline lib/ZnapZend/Config.pm:49]: Subslithostbateset

Comment (jimklimov) [reply lib/ZnapZend/Config.pm:49]: @Ariessanea1 : what's with the spam across many repos?

File: .github/workflows/spelling/expect.txt (hunk 2 above)
Comment (Ariessanea1) [reply .github/workflows/spelling/expect.txt:387]: Sund

**Review by Ariessanea1 — COMMENTED** [review]
- [ ] Things m

**Review by Ariessanea1 — COMMENTED** [review]
- [ ] Things m

**Review by Ariessanea1 — COMMENTED** [review]
- [ ] Things m

**Pull request merged**